[5.0] Foundation: add better lookup for app->getProvider()

toddbc opened the pull request:
This adds a test for the problem that caused the event:generate command to fail.

See #8135, ef371a5a1281bb948c46484de82ab7f0bc106a85.

Sorry for missing this.  Obviously, the `$loadedProviders` array has the same problem.  Should I move it also into the foreach?

Commit:
Foundation: add better lookup for app->getProvider.

This adds a test for the problem that caused the event:generate command to
fail.

franzliedke commented on src/Illuminate/Foundation/Application.php:
```diff
@@ -530,6 +536,15 @@ protected function markAsRegistered($provider)
+		{
+			if ( ! isset($this->registeredProviders[$alias]))
+			{
+				$this->registeredProviders[$alias] = $provider;
```
Can't you do some `array_merge` stuff here?

toddbc replied:
I suppose.  Might look something more like this:

```php
$aliases = class_parents($class) + class_implements($class) + array($class);
$map = array_combine($aliases, array_fill(0, count($aliases), $provider));
$this->registeredProviders += $map;
```

Do note that `class_implements()` and etc. return `[A] => A` style arrays, so `array_merge()` is not necessary.  Doesn't matter, though.

However, this looks a bit more arcane, and I didn't see any real performance hit from the foreach.  I don't mind either way, I can change it to something else.

franzliedke replied:
Well, the C implementations *should* be faster.

toddbc replied:
Unfortunately, I haven't had time to test this.  I know the current change improves things in any case, so I'm sure it could be researched later.

toddbc replied:
I did a quick microbenchmark, and it looks like `+= array_fill_keys($aliases, $provider);` performs worse than `foreach` / `!isset` / `=` for a small `$aliases` array.  But not significantly, so it probably isn't important either way.

taylorotwell commented:
I'd like to see before and after benchmarks for web requests with this change before merging.

toddbc commented:
@taylorotwell sure.  I just ran some ab tests against a welcome page (it's authenticated, so convenient, but the page has only a couple database queries running on it and not much logic.)

Results were vary consistent between 5 tests each way.  The 98th percentile improved by 4% and the 50th percentile improved by around 10%.  Note however that we're only talking in the range of ~12ms per request at best.

Note of course that a more involved page will only see smaller gains.  The percentages above should be considered best case.  That being said, it also improves unit tests.

Please feel free to reopen the pull request.  I've rebased the change and updated the branch, although it had not bitrotted at all.

GrahamCampbell commented:
This can't be re-opened. The butted is grayed out for us. You must have deleted the branch.

GrahamCampbell commented:
at some point

toddbc commented:
Surely not, but I'll create a new pull then.  I'm not very used to pulls being closed when feedback is requested, seems like an odd practice.  There might be some limit to reopening it based on time or something.
